### PR TITLE
Don't default to object detail for single result questions

### LINKED ISSUE #21873: Can't exit object detail modal when there's only one result
**Describe the bug**
Object detail modal defaults top open when there's only one result

**To Reproduce**
1. Filter a table by a single ID (primary key)
2. See that the record opens in object detail view modal
3. See that you can't close the modal

**Expected behavior**
1. should probably open single results in table view first, and be able to open and close the modal at will.

**Screenshots**
Screenshot from 2022-04-21 10-25-19

**Severity**

Quite painful for users to have to use the back button, etc to edit filters.

### PR DESCRIPTION
**Pull request by iethree**
Previously, any query that had a single result would go directly to the object detail view for that single result, and would not, in fact, allow the user to see the table view for that result at all. With the moving of object detail's primary view to a modal - this made that modal inescapable :cry: 

This changes that behavior to always show a single result in a table and require a user to click into that result to see the object detail view.

To test:
1. browse the orders table
2. filter > order id = 1
3. see single result in table
4. you can click into and out of object detail modal for that record

Resolves #21873 

Screenshot from 2022-04-21 15-00-32

### CODE REVIEW

**Comment by iethree** [comment]
@mazameli Can you confirm that there isn't any place where we want to _always_ send a user to the object detail view - except when the object has been specifically "zoomed" ?  

This change will not prevent direct-linking to a detail view, but it will make it so the user can always exit detail view back to table view - which was not the case before.

**Comment by mazameli** [comment]
@iethree none come to mind (except for directly linking to the modal or fullpage view of a record, which you already addressed).

**Review by alxnddr — APPROVED** [review]
Looks good, works as expected

File: frontend/test/metabase/scenarios/question/new.cy.spec.js
Comment (iethree) [inline frontend/test/metabase/scenarios/question/new.cy.spec.js:236]: Setting quantity as a primary key seems like a nonsensical test to me, but I don't want to stop testing something if this is somehow important. 

@kulyk, it looks like you touched this test somewhat recently, can you see any reason we want to keep that part of the test?

Comment (kulyk) [reply frontend/test/metabase/scenarios/question/new.cy.spec.js:236]: @iethree this mimics a table with composed primary keys, ie when there are > 1 columns needed to uniquely identify a row, so I believe it should actually stay here

Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/question/new.cy.spec.js:236]: @kulyk is right. This was the essential part of this reproduction/issue.

Please, see: https://github.com/metabase/metabase/issues/13717

Comment (kulyk) [reply frontend/test/metabase/scenarios/question/new.cy.spec.js:236]: @iethree as a possible solution, AFAIR there's an idea to add a full-screen mode for object detail (not only a modal). Would it make sense to always open full-screen object detail for queries returning a single row without making it possible to open the modal?

**Review by nemanjaglumac — DISMISSED** [review]
Blocking this just so the comments about the repro don't get lost.
This has nothing to do with the actual FE code, and I'll approve once the changes @kulyk and I requested are addressed.

**Review by mazameli — APPROVED** [review]
This is the behavior I'd expect.

File: frontend/test/metabase/scenarios/question/new.cy.spec.js
Comment (iethree) [reply frontend/test/metabase/scenarios/question/new.cy.spec.js:236]: Thanks, that's helpful context, and a good test. I'll need to rework the underlying code slightly to correctly account for composite keys.

Comment (iethree) [reply frontend/test/metabase/scenarios/question/new.cy.spec.js]: 1. I have restored the API call in this test to make it a valid test of composite keys. I have changed the test assertions and description to accurately test the current filter behavior of multiple primary keys.
2. Instead of hacking-in a new way to handle composite keys, I've essentially disabled object detail view for items with composite keys in this branch. I think this can be simply and elegantly handled with currently-planned work around object detail view: namely: making it viewable independent of whether a question has a visible primary key (i.e. just show a row). 
3. This introduces a temporary regression (no object detail view for composite keys), but I think that makes sense with the direction this feature is going (ie. object detail is always reached by "zooming in"), which is not currently technically possible, because the current zoom logic relies on primary keys.

Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/question/new.cy.spec.js]: I've noticed one difference with Object detail window now, compared to the 42.4, for example.
If you click on a primary key in a joined question it now displays ALL fields (primary and joined). Is this a desired behavior?

Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/question/new.cy.spec.js]: @iethree I've just realized one thing, and it was my mistake when I originally wrote the repro. If you read @flamber 's repro steps, you'll notice he says to set `ORDERS.ID` type to "No semantic type" in the Data model.

For whatever reason, I overlooked that part originally.
If you try the repro following the exact steps from the issue, you'll notice it still opens the object detail modal and the results are wrong. Please take a look at this video:
https://www.loom.com/share/579f38c4ee3b48b4b3d1f3876d9d2a36

File: frontend/test/metabase/scenarios/question/new.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/question/new.cy.spec.js:261]: Let's please add `cy.wait("@dataset");` after each of these clicks.

[L:243](https://github.com/metabase/metabase/pull/21895/files#diff-7ea202f142e9881a22ed40fdf5aa77b70cc3a58b287d8674cb6d94fbd486cb4aR243)
[L:260](https://github.com/metabase/metabase/pull/21895/files#diff-7ea202f142e9881a22ed40fdf5aa77b70cc3a58b287d8674cb6d94fbd486cb4aR260)

Otherwise I'm pretty confident we'll see flakes sooner or later.

Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/question/new.cy.spec.js:261]: Btw, you don't have to intercept the `/api/dataset`. That's already a part of the `openOrdersTable()` helper.

File: frontend/test/metabase/scenarios/question/reproductions/17514-ui-overlay.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/question/reproductions/17514-ui-overlay.js:145]: @iethree please revert this change.
`visualize()` helper already waits for `@dataset` so this will make things only worse.

Looking at the screenshot for the failed test, it doesn't seem like a flake but a legitimate failure because object detail window now appears.
image

In order to investigate failures in the CI, please go to the failed run and then [click on Artifacts tab](https://app.circleci.com/pipelines/github/metabase/metabase/34791/workflows/d81f915f-30b0-470f-807d-27a77be2a4ef/jobs/1807973/artifacts). You'll see both screenshots and videos there.

image

Comment (iethree) [reply frontend/test/metabase/scenarios/question/reproductions/17514-ui-overlay.js:145]: The failure is because it's trying to click on `110.93`, but it actually clicks on the id column because it clicks before the UI is fully loaded and there's a small visual shift.  Any suggestions on how to wait for that UI load more accurately?

File: frontend/test/metabase/scenarios/question/new.cy.spec.js
Comment (iethree) [reply frontend/test/metabase/scenarios/question/new.cy.spec.js]: @nemanjaglumac took a look at your video, and it makes sense to me. The issue is "zooming" in on a single record currently requires a unique primary key. if you set quantity as the primary key (or have a composite primary key), what it currently does is show you the last instance of that key in the current dataset.

This is what i was referring to in terms of a temporary regression until we can enhance object detail "zooming" to support rows without unique primary keys.

File: frontend/test/metabase/scenarios/question/reproductions/17514-ui-overlay.js
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/question/reproductions/17514-ui-overlay.js:145]: Well, that's a bummer then! :(
Up until fairly recently it was enough to wait for the `response` (in this case `POST /api/dataset`), but then we started having more and more these visual glitches and it takes browser a bit more to actually render results.

To make things worse, this test already implements a "second level" guard by additionally making sure that the table header is rendered:
`cy.findByTextEnsureVisible("Subtotal");`

I don't have any other ideas other than try to make sure 110.93 is visible before clicking on it.
`cy.findByTextEnsureVisible("110.93");`

That's still not going to be a guarantee, but it's the best we can do at the moment.

Comment (iethree) [reply frontend/test/metabase/scenarios/question/reproductions/17514-ui-overlay.js]: @nemanjaglumac take a look at the latest commit. What we're testing here is whether the table UI is clickable at all - so instead of having the assertion on a table cell, I've updated the assertion to be on the table header which seems less prone to shifting.

Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/question/reproductions/17514-ui-overlay.js]: Works for me.
Great thinking. Thanks

**Comment by iethree** [comment]
returning this to draft until https://github.com/metabase/metabase/pull/22190 and related issues get sorted

File: frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx
Comment (iethree) [inline frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx:21]: @kulyk I think this is a simpler way to accomplish preserving dashboard filters when drilling through primary keys. Let me know if you think I'm missing something.

**Review by kulyk — APPROVED** [review]

File: frontend/src/metabase/query_builder/selectors.js
Comment (iethree) [reply frontend/src/metabase/query_builder/selectors.js:483]: I think we should leave it for now until we hash out exactly how full-page object detail view will work

see epic: https://github.com/metabase/metabase/issues/22119

File: frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx
Comment (iethree) [reply frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx:88]: 1) I don't think so, since currently we're always fitlering to a single row for FK drills
2) I think composite key edge cases are better handled in this PR: https://github.com/metabase/metabase/pull/22398, which greatly simplifies viewing them

File: frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx
Comment (iethree) [reply frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx:21]: my misgiving about the getExtraData approach is that it obscures what is going on down the chain and makes things hard to search for. I think we can the best of both worlds though by explicitly naming which extra props we're passing. I'll make the update.

File: frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx
Comment (kulyk) [reply frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx:88]: Should we use the `getExtraData` approach then as it used to be until the next PR is merged?

**Pull request merged**
